metadata/src/main/java/org/apache/kafka/image/MetadataDelta.java

```diff
@@ -253,6 +263,14 @@ public void replay(RemoveTopicRecord record) {
 
     public void replay(FeatureLevelRecord record) {
         getOrCreateFeaturesDelta().replay(record);
+        featuresDelta.metadataVersionChange().ifPresent(__ -> {
+            // If any feature flags change, need to trigger components to possibly handle the new flags
```

Comment:
hmm, you forgot AclsDelta here :)

I think we need a separate function to handle changes in `metadataVersion`, not just `getOrCreate*Delta`.

After all, changes in `metadataVersion` could have the effect of deleting a bunch of stuff (downgrade to version X without the stuff). And this is an immediate effect: as soon as the downgrade happens, boom. The data is gone.

Whereas `getOrCreate*Delta` is a no-op if the delta object already exists.